### ISSUE TITLE
Unable to 'replicate' ReadFish testing results

### ISSUE THREAD

**Issue by sirselim** [LooseLab/readfish#142]
Hi team, 

Firstly thanks for all your hard work and the awesome tool.

For the life of me I can't figure out what I am doing wrong, I'm picking it's me being stupid at some stage but have tried troubleshooting past the point where I probably should have given up and posted here...

I'm trying to get ReadFish running on my workstation (Lenovo P920, running latest Debian). Specs:

- 2x12 core Intel Xeon Gold 5118 (48 threads @ 3.2GHz)
- 256GB RAM
- plenty of fast SSD
- Nvidia Titan RTX (24GB RAM)

MinKNOW is latest (21.02.1, core 4.2.5) with Guppy 4.3.4.

I built the dev branch of ReadFish and acknowledge that there is no pyguppy-client available for 4.3.4, so am using 4.4.2 (if this is going to be a problem I could look at downgrading everything, but it appears to be 'working'/running).

When I kick of `readfish unblock-all` things appear to be happening, see below:
![image](https://user-images.githubusercontent.com/5932864/111932411-363c4300-8b22-11eb-9871-b576a09d588a.png)

However it really doesn't seem to be responding in the manner in which I would expect, or that the testing page suggests. Below is the output from ~20mins of sequencing supposedly unblocking all reads:  
![image](https://user-images.githubusercontent.com/5932864/111932295-03924a80-8b22-11eb-9535-6a5dca5f0f53.png)

...and a zoom in of that:
![image](https://user-images.githubusercontent.com/5932864/111932353-19a00b00-8b22-11eb-8d77-c8ab9cf181be.png)

So while there is the peak at ~500bp it's by no means the majority of the sequence data generated.

I've tried various tweaks, including commenting out adapter scaling and playing with the throttle. I've also been tweaking the basecalling parameters for the guppy server as the Titan RTX is no slouch but isn't at the same level as the GV100 in the GridION.

Another thing that might be useful to note is that I can't seem to get the built in ReadUntil running through MinKNOW correctly either. It runs and generates a log, but only a small proportion of the reads are actually unblocked (I've tried various iterations on the references and 'targets' etc).

I'm beginning to wonder if this is maybe something that is tied to running in simulation/playback mode, but if so I don't understand why I can't at least replicate the ReadFish results with the same dataset. I would love to test this with a real experiment but unfortunately at the moment our lab is very COVID centric and it will be a little while before we can start research runs again. 

Any advice would be very much appreciated. :) Please let me know if I need to supply further information.

**Comment by alexomics** [comment]
From your description it seems like you've set everything up as we'd expect to work. Regarding pyguppy-client-lib, that should be fine, generally the python library should be less than or equal to the guppy server version. 

How have you setup up the environment, i.e are you using conda or another python source? What python version? 

Is this running from our bulk FAST5 file (PLSP57501_20170308_FNFAF14035_MN16458_sequencing_run_NOTT_Hum_wh1rs2_60428.fast5)?

Finally have you changed the `break_reads_after_seconds` field in the sequencing TOML config?

**Comment by sirselim** [comment]
I've just set up Readfish in a python virtual env, using python 3.7.6 (ish, definitely 3.7.X - can't check at the moment).

Yep, this is running the same bulk file that you link in the testing section of the wiki.

I have changed that parameter in the 'app_conf' TOML to 0.4.

**Comment by alexomics** [comment]
Okay, let me have a think. I'll try to get some troubleshooting steps together

**Comment by mattloose** [comment]
OK - this is frustrating I am sure!

For the histogram of read lengths that you show above can you show the histogram with the "split by read end reason" selected?

**Comment by mattloose** [comment]
![image](https://user-images.githubusercontent.com/8202484/111984785-bc955c80-8b03-11eb-850f-f3d753d60a97.png)
Just confirming for the record that this is what we would expect you to see - we will obviously try and help you get to the bottom of this!

**Comment by mattloose** [comment]
![image](https://user-images.githubusercontent.com/8202484/111986482-ebaccd80-8b05-11eb-8a52-bd56b563d766.png)
This is the same thing about 20 minutes in... so something here that needs some tweaking.

**Comment by sirselim** [comment]
Thanks Alex and Matt for delving into this. 

I nearly added a 'split by read end reason' with the original post, sorry! I'll be in the office later this afternoon so will grab it for you. 

I know this is crude, but the below doodle indicates the purple 'data service unblock mux change'. 

![IMG_20210323_072226.png](https://user-images.githubusercontent.com/5932864/112039211-ad1c1f00-8ba8-11eb-8700-9822d934feff.png)

**Comment by sirselim** [comment]
Actually, looking at your screenshots, are you not running basecalling during that run?

**Comment by mattloose** [comment]
You are correct - on that we weren't. In our hands it doesn't make a difference to the performance (I'll post a screenshot in a bit to show that).

But this might be worth you checking. What happens if you run the unblock all script on playback without live basecalling switched on? Does that change the behaviour at all?

**Comment by sirselim** [comment]
Yeah I haven't tried without live basecalling, something else to test this afternoon.

I am wondering if I have a bigger issue as the Minknow adaptive sampling is not playing ball either.

**Comment by alexomics** [comment]
We haven't been able to re-create this on the GridION, which is a little frustrating. I do have some more questions about your setup:
 - Are you running GPU base calling for live base calling?
 - If so, what kind of resources is it setup to use?
 - Have you tried a 'targets' run yet? It'd be interesting to see if this behaviour is restricted to just unblock-all

It might also be worth checking out the log files that minknow writes, these are usually under `/var/log/minknow/<position>/bream-<x>.txt`

Another small test it might be worth trying is setting `--throttle` to either `0.4` or `1.0`. From your screenshot of the console logs it looks like each successive batch is >1s apart where we would expect them to be closer to 0.4s apart

**Comment by mattloose** [comment]
![image](https://user-images.githubusercontent.com/8202484/112048339-33ebe000-8b46-11eb-9845-c4ea82ccff67.png)
Just to add - this is a longer "unblock all" with basecalling switched on. As you can see it still "works" (though personally I'd like to see it a bit cleaner than this....).

It is worth pointing out that it is pretty clean by read count (rather than read length!). 
![image](https://user-images.githubusercontent.com/8202484/112048522-6ac1f600-8b46-11eb-9fe5-8767a22fb8f5.png)

**Comment by sirselim** [comment]
Thanks both.

@alexomics some answers to your questions:

- yep, I've been running live gpu basecalling during all testing. I'll try running without and report back.

- in terms of Guppy parameters, I believe something like:          
  - 48 gpu_runners_per_device
  - 256 chunks_per_runner
  - 2000 chunk_size
But I've played with many and varied parameters here.

- I'll try a target's run later as well

- cool, I can report back with some log files

- I've played with the --throttle setting a little and didn't notice any difference, but I'll run some more tests

@mattloose - thinking about it now, this is the same thing I was seeing when I was testing Readfish on the Jetson Xavier AGX. I put it down to not enough hardware headroom on that device but maybe something else is going on?

**Comment by mattloose** [comment]
Yes - I wonder if you are seeing the same thing. My suspicion is though that you will find the simple unblock-all script doesn't work here - and solving this is key to making sure you will be able to run adaptive sampling (any of it's various implementations) - so lets see if we can work out what the issue is!

**Comment by sirselim** [comment]
OK, so I have some updates.

Running without live GPU basecalling turned on seems to help out a lot. Here is a snap about 6 minutes into a run:

![image](https://user-images.githubusercontent.com/5932864/112075049-90014380-8bdc-11eb-9f2a-3002972592cb.png)

However it is odd that I'm not getting as 'clean' a response as you @mattloose. Here is the same run now 21 minutes in:

![image](https://user-images.githubusercontent.com/5932864/112075916-487bb700-8bde-11eb-8afd-fbc91cbe0f9f.png)

"Zoomed" in on the small read peak:

![image](https://user-images.githubusercontent.com/5932864/112075978-6517ef00-8bde-11eb-91fe-eb014ea9e542.png)

... and just looking at the counts, so similar to the graph for counts you showed this is looking pretty clean:

![image](https://user-images.githubusercontent.com/5932864/112076019-7a8d1900-8bde-11eb-8eb4-0f621bf346f7.png)
 
@alexomics - here is the bream log file from around the 20 minute mark if that is at all useful ([bream-0.txt](https://github.com/LooseLab/readfish/files/6186083/bream-0.txt))

I'll now test a targets run and report back with those results soon.

**Comment by sirselim** [comment]
A slight update on this.

I have gone back to the testing I was doing a few months ago using ReadFish on the Jetson Xavier AGX. I had given up back then thinking that the hardware just wasn't up to it, but after this recent bout of troubleshooting I thought I'd have a further play. So long story short, I set up a simulation run without GPU enabled live basecalling and unblocks appear to be much more prevalent than they were previously, nowhere near 'clean' in terms of other/longer reads coming through, but a distinct peak at the ~500bp mark. So that's quite interesting.

On the back of this I tried to perform a target run, but it looks like I'm getting the same error with ReadFish - which I am putting down to mismatching pyguppy-client-libs. I'm probably going to see if ONT can release a version in line with Guppy 4.3.4, which is what is currently being used with the latest MinKNOW.

All of this has got me thinking, and reflecting on past conversations we had @mattloose, about CPU requirements. For ReadFish are there any set CPU single core 'speed' requirements, or to rephrase, have you noticed performance is much better when you have faster cores? The ARM CPU cores on the Xavier top out at ~2.3GHz, and while the CPUs in my workstation are core count plentiful they only ramp up to ~3.2GHz (turbo frequency). I'd be interested to know what you guys have been testing on in house, and even what sort of CPU is inside the GridION (edit: I forgot I have access to a couple now! It has an Intel Core i7-7700, which can turbo to 4.2GHz). 

So the CPU in the GridION is able to turbo cores to clock speeds 1GHz higher than the CPUs in my workstation, and these are ~2GHz faster than what is available on the Xavier. Is this possibly influencing what we are seeing, or a red herring? I can't readily think of any machines I have to hand that I can easily get the ONT/ReadFish stack up and running on to test this out. I could try on my personal Dell Precision laptop, it doesn't have a compatible dedicated GPU but I shouldn't need that just to test `unblock_all` speed right?

**Comment by mattloose** [comment]
So - worth noting that ONT have now released a beta version of read until that is compatible with Mk1C - it might be worth you looking at that in your Jetson Xavier AGC context.

I think the unblock-all performance is not linked to GPU nor should it be. The fundamental issue here is that unblock-all is as light-weight as we can make it and should always work... I think I've annoyed enough people over the years with my obsession with this - but at least in my hands since read until became available it has only worked when that unblock all simple routine has performed well. I have also had cases where we've "lost" performance over time. I will ask @alexomics to let me know the cpu of the "worst" system we've tried on.

**Comment by sirselim** [comment]
OK, so a lot has happened in the time since reporting this 'issue'. I'm not sure I'm any closer to having a solution but I do have a few more pieces to add to the puzzle. These come from my work getting adaptive sampling running 'natively' on the Jetson Xavier AGX and Xavier NX. I don't want to dump a massive wall of text here, but I have written up my findings and added them to my Jetson Nanopore repo - if you are interested in having a read here is the [link](https://github.com/sirselim/jetson_nanopore_sequencing/blob/main/xav_adaptive_sampling.md).

It essentially boils down to the fact that when live GPU basecalling is turned off on these devices they will happily (and very nicely) perform adaptive sampling as implemented in MinKNOW - though the NX requires a little bit of 'tweaking' before this can happen. What is even more interesting is that with the Xavier AGX you can lunch an independent instance of Guppy to basecall the data that is being generated and it will happily run everything at once. But if you try to do it with live GPU basecalling + adaptive sampling there is a very large lag in making the unblock decision.

@mattloose - if you get a chance to read through this I would be really interested in your take.
